### PR TITLE
[DO NOT MERGE] Fix "rm -r" function and normalize path string with extra '/' characters.

### PR DESCRIPTION
**Pull request by amogh-samsung**
- Fix smartfs_finddirentry() returns number of bytes read from flash when path ends in extra '/' character due to retention of last
  assigned value to return variable.
- Repeated '/' in path should be ignored to align with POSIX standard.
- For Eg: A) paths = "/mnt/test1" and = "/mnt/test1/" and B )paths = "/mnt/test1" and "/mnt//test1" should be considered
  same while looking for the entry respectively..
- Fix behavior of "rm -r", should only delete directory paths

Signed-off-by: Amogh Hassija <a.hassija@samsung.com>

### CODE REVIEW

File: apps/system/utils/fscmd.c [hunk 1]
```diff
@@ -1084,10 +1084,26 @@ static int tash_rm(int argc, char **args)
 {
 	char *fullpath;
 	int ret = OK;
+	struct stat st;
 
 	if (args[2]) {
 		fullpath = get_fullpath(args[2]);
 		if (strncmp(args[1], "-r", strlen(args[1])) == 0) {
+			/* The request is to delete an entire directory recursively
+			   Check if the entry requested is not a direcotyr and report
```
Comment (dineshdas-samsung) [inline apps/system/utils/fscmd.c:1093]: Can you please check the spelling of directory ?

Comment (amogh-samsung) [reply apps/system/utils/fscmd.c]: Rectified, thanks

File: apps/system/utils/fscmd.c [hunk 2]
```diff
@@ -1097,12 +1113,12 @@ static int tash_rm(int argc, char **args)
...
 			FSCMD_OUTPUT(CMD_FAILED, args[0], "unlink");
-			fscmd_free(fullpath);
-			return ret;
+			goto errout;
 		}
 		FSCMD_OUTPUT("%s deleted\n", fullpath);
 	}
 
+errout:
 	if (fullpath) {
```
[the remark this answers is not in the document]
Comment (amogh-samsung) [reply apps/system/utils/fscmd.c]: Thanks for pointing that out :)

**Review by deepaksrma — COMMENTED** [review]
LGTM

File: apps/system/utils/fscmd.c [hunk 3]
```diff
@@ -1084,10 +1084,26 @@ static int tash_rm(int argc, char **args)
...
 
 	if (args[2]) {
 		fullpath = get_fullpath(args[2]);
 		if (strncmp(args[1], "-r", strlen(args[1])) == 0) {
+			/* The request is to delete an entire directory recursively
+			   Check if the entry requested is not a directory and report
+			 */
+			if (stat(fullpath, &st) < 0) {
+				FSCMD_OUTPUT("stat failed with %s\n", fullpath);
+				ret = ERROR;
```
Comment (jeongchanKim) [inline apps/system/utils/fscmd.c:1097]: In `stat`, it sets the errno before returning failure case. So I think we'd better get the errno and set the `ret` as errno.

File: apps/system/utils/fscmd.c [hunk 4]
```diff
@@ -1097,15 +1113,13 @@ static int tash_rm(int argc, char **args)
 		ret = unlink(fullpath);
 		if (ret != OK) {
 			FSCMD_OUTPUT(CMD_FAILED, args[0], "unlink");
-			fscmd_free(fullpath);
-			return ret;
+			goto errout;
```
Comment (jeongchanKim) [inline apps/system/utils/fscmd.c:1116]: `unlink` also sets the errno on failure case. How about setting the `ret` as -errno?

File: apps/system/utils/fscmd.c [hunk 5]
```diff
@@ -1084,10 +1084,26 @@ static int tash_rm(int argc, char **args)
...
+			/* The request is to delete an entire directory recursively
+			   Check if the entry requested is not a directory and report
+			 */
+			if (stat(fullpath, &st) < 0) {
+				FSCMD_OUTPUT("stat failed with %s\n", fullpath);
+				ret = ERROR;
+				goto errout;
+			}
+
+			if (!S_ISDIR(st.st_mode)) {
```
Comment (jeongchanKim) [inline apps/system/utils/fscmd.c:1101]: In this case, is `unlink` not needed?
In `delete_entry()`, there is an unlink with same case.

Comment (amogh-samsung) [reply apps/system/utils/fscmd.c:1101]: This case checks if the path given along with rm -r does not represent a directory, then the command is invalid.

File: apps/system/utils/fscmd.c (hunk 3 above)
Comment (amogh-samsung) [reply apps/system/utils/fscmd.c]: Yes, updated

File: apps/system/utils/fscmd.c (hunk 4 above)
Comment (amogh-samsung) [reply apps/system/utils/fscmd.c:1118]: Updated, thanks

File: apps/system/utils/fscmd.c [hunk 6]
```diff
@@ -1097,15 +1114,14 @@ static int tash_rm(int argc, char **args)
...
+			ret = -errno;
+			goto errout;
 		}
 		FSCMD_OUTPUT("%s deleted\n", fullpath);
 	}
 
-	if (fullpath) {
-		fscmd_free(fullpath);
-	}
+errout:
```
Comment (thapav) [inline apps/system/utils/fscmd.c:1123]: Is the if(fullpath) check not necessary anymore since its been deleted ?

Comment (amogh-samsung) [reply apps/system/utils/fscmd.c:1123]: Old comment on the same line.
fscmd_free(path) checks if path is NULL internally.

**Review by jeongchanKim — DISMISSED** [review]
LGTM

**Review by thapav — DISMISSED** [review]
LGTM

File: apps/system/utils/fscmd.c [hunk 7]
```diff
@@ -1084,10 +1085,26 @@ static int tash_rm(int argc, char **args)
...
 	int ret = OK;
+	struct stat st;
 
 	if (args[2]) {
 		fullpath = get_fullpath(args[2]);
 		if (strncmp(args[1], "-r", strlen(args[1])) == 0) {
+			/* The request is to delete an entire directory recursively
+			   Check if the entry requested is not a directory and report
+			 */
+			if (stat(fullpath, &st) < 0) {
```
Comment (Taejun-Kwon) [inline apps/system/utils/fscmd.c:1096]: Below check code already duplicated, delete_entry also checking this, they are almost same.

Comment (amogh-samsung) [reply apps/system/utils/fscmd.c:1096]: `delete_entry()` is being called recursively for all subentries in the directory.
Sometimes the entry is a file and sometimes a directory, but we cannot check if it is the parent entry for which the function is invoked.
So, despite `stat` already being invoked in `delete_entry()`, I had to add a check here...

**Comment by amogh-samsung** [comment]
This PR is similar to #4922 
Both offer a solution to the same bugs with a different version of the changes.
4922 includes changes to fs layer majorly to handle path discrepancies.
4928 includes changes to vfs layer majorly to handle path discrepancies.

File: os/fs/vfs/fs_open.c [hunk 8]
```diff
@@ -137,7 +141,15 @@ int vopen(FAR const char *path, int oflags, va_list ap)
 
 	/* Get an inode for this file */
 
-	inode = inode_find(path, &relpath);
+	if (strncmp(path, "/", 1) != 0) {
+		tmp = (char *)kmm_malloc(strlen(path));
```
Comment (dineshdas-samsung) [inline os/fs/vfs/fs_open.c:145]: What if tmp is NULL ? Can we please handle this case ?

File: os/fs/vfs/fs_rmdir.c [hunk 9]
```diff
@@ -111,13 +113,19 @@ int rmdir(FAR const char *pathname)
...
+	const char *norm_path = NULL;
+	char *tmp = NULL;
 	int errcode;
 
 	/* Get an inode for this file.  inode_find() automatically increments the
 	 * reference count on the inode if one is found.
 	 */
 
-	inode = inode_find(pathname, &relpath);
+	tmp = (char *)kmm_malloc(strlen(pathname));
```
Comment (dineshdas-samsung) [inline os/fs/vfs/fs_rmdir.c:124]: Same as above.

File: os/fs/vfs/fs_norm_path.c [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ ****************************************************************************/
+
+void normalize_path(const char *path, char **ret_path)
+{
+	int i;
+	int j = 1;
+	char *tmp = *ret_path;
+
+	/* Copy the first character anyways */
+	tmp[0] = path[0];
```
Comment (dineshdas-samsung) [inline os/fs/vfs/fs_norm_path.c:60]: Can you please check ?
Since, there was also no check while calling  normalize_path(path, &tmp) for tmp==NULL;

**Comment by amogh-samsung** [comment]
Under reconsideration

**Pull request closed without merging**